cpp/src/arrow/compute/kernels/cast.cc

```diff
@@ -396,21 +396,34 @@ struct CastFunctor<Date64Type, TimestampType> {
     ShiftTime<int64_t, int64_t>(ctx, options, conversion.first, conversion.second, input,
                                 output);
 
-    internal::BitmapReader bit_reader(input.buffers[0]->data(), input.offset,
-                                      input.length);
+    if (input.null_count != 0) {
+      internal::BitmapReader bit_reader(input.buffers[0]->data(), input.offset,
+                                        input.length);
 
-    // Ensure that intraday milliseconds have been zeroed out
-    auto out_data = GetMutableValues<int64_t>(output, 1);
-    for (int64_t i = 0; i < input.length; ++i) {
-      const int64_t remainder = out_data[i] % kMillisecondsInDay;
-      if (ARROW_PREDICT_FALSE(!options.allow_time_truncate && bit_reader.IsSet() &&
-                              remainder > 0)) {
-        ctx->SetStatus(
-            Status::Invalid("Timestamp value had non-zero intraday milliseconds"));
-        break;
+      // Ensure that intraday milliseconds have been zeroed out
+      auto out_data = GetMutableValues<int64_t>(output, 1);
+      for (int64_t i = 0; i < input.length; ++i) {
+        const int64_t remainder = out_data[i] % kMillisecondsInDay;
+        if (ARROW_PREDICT_FALSE(!options.allow_time_truncate && bit_reader.IsSet() &&
+                                remainder > 0)) {
+          ctx->SetStatus(
+              Status::Invalid("Timestamp value had non-zero intraday milliseconds"));
+          break;
+        }
+        out_data[i] -= remainder;
+        bit_reader.Next();
+      }
+    } else {
+      auto out_data = GetMutableValues<int64_t>(output, 1);
+      for (int64_t i = 0; i < input.length; ++i) {
+        const int64_t remainder = out_data[i] % kMillisecondsInDay;
+        if (ARROW_PREDICT_FALSE(!options.allow_time_truncate && remainder > 0)) {
```

Comment:
Looks like the `options.allow_time_truncate` test and also the `out_data` declaration could be factored out.

Reply:
@pitrou I'm not sure how should I factor out `options.allow_time_truncate` check. Use a single loop and check for the `null_count` inside?

Reply:
`options.allow_time_truncate` is a constant accross this whole piece of code, so just add a higher-level `if` statement around all this.

Reply:
Doesn't the first value encountered with time part trigger the error - which has to be checked inside the loop?

Reply:
What I mean is that you can skip the whole thing is `options.allow_time_truncate` is true (the compiler might do the optimization for us, but still).

Reply:
I might misunderstand, but:

```python
# with allow_time_truncate
[
    '2018-05-10T00:00:00',
    '2018-05-11T00:00:00',
    '2018-05-12T10:24:01',
]  # OK

# without allow_time_truncate
[
    '2018-05-10T00:00:00',
    '2018-05-11T00:00:00',
    '2018-05-12T10:24:01',  # <- fails here
]  

# with allow_time_truncate
[
    '2018-05-10T00:00:00',
    '2018-05-11T00:00:00',
    '2018-05-12T00:00:00',
]  # OK

# without allow_time_truncate
[
    '2018-05-10T00:00:00',
    '2018-05-11T00:00:00',
    '2018-05-12T00:00:00',
]  # OK - this would fail if I test outside the loop


```

Reply:
Sure, but don't we need another branch then to handle when time truncation is allowed?

Reply:
No problem :) I'm still learning arrow.